feat: shutdown in reverse dependency order

anmonteiro opened the pull request:
processes that depend on others to start may also need to finish before them and perform some cleanup. This change makes sure that's the case.

fixes #145

F1bonacc1 reviewed (changes requested):
Hey @anmonteiro,

Thanks for the PR, there are definitely scenarios where this feature is required.
I am a bit concerned that for cases where this is not required, this will make the shutdown sequence much longer.

Since it's a behavior change I don't think we can make it the new default.
A global `is_ordered_shutdown` flag (default `false`) will do the trick.

Will you be able to add a few `ShutDownProject()` tests to `system_test.go` to make sure that it works as expected?

P.S.
Loved the algorithm you used to wait for the dependencies' completion without shutting them down. Nice!

anmonteiro commented:
@F1bonacc1 thanks for the early draft review. I believed I addressed all your review comments in the latest commit.

Just missing the tests, which I'm working on right now.

anmonteiro commented:
added a skeleton of a test case shutting down the runner and checking for 0 running processes, but I'm not sure how to check that the processes are actually stopped in the order we expect.

thenonameguy commented on fixtures-code/process-compose-shutdown-inorder.yaml:
```diff
@@ -0,0 +1,31 @@
+    isDaemon: true
+    shutdown:
+      command: |
+        echo "C: exiting"
```
You could write these to a file, then assert at the end for the contents to be ordered backwards.
```suggestion
        echo "C: exiting" >> shutdown-inorder.log
        sleep 1
```

F1bonacc1 replied:
Another approach:
https://github.com/F1bonacc1/process-compose/commit/9ac79177aeed4fb471faa644fbe4ccf9ad9a7360

anmonteiro replied:
Nice, thanks. I cherry-picked it to this branch.

anmonteiro commented:
Not sure why it fails in CI, `make test` succeeds locally.

F1bonacc1 commented on src/app/system_test.go:
```diff
@@ -406,3 +410,93 @@ func TestSystem_TestProcListToRun(t *testing.T) {
+			return
+		}
+		runningProcesses := 0
+		for _, proc := range runner.runningProcesses {
```
Race condition here.
You can replace this block with `runner.GetProcessesState()` (plural)

F1bonacc1 replied:
Fixed here:
https://github.com/F1bonacc1/process-compose/commit/27c3ab97973b78499a38d50fbd41d1d4514224eb

F1bonacc1 commented:
Race condition

> Not sure why it fails in CI, `make test` succeeds locally.

Race condition on `runner.runningProcesses`.
Try `make testrace`
Left you a comment at the proper line.

anmonteiro commented:
I actually believe it might be the log rendering using ANSI escape codes but not the wanted string slice.

anmonteiro commented:
pushed a fix using `GetProcessesState()` and adding `NoColor: true` to the logger configuration.

F1bonacc1 commented:
Looks good! Thanks!

F1bonacc1 commented:
Will you be able to squash?

anmonteiro commented:
> Will you be able to squash?

Absolutely. Done.
